README.md
# 🛍️ Store 🛍️


#### This is a project made during Alura's Clojure course

---

## 📌 About the project
A program that makes simple operations of a store, with the objective of fixing the course content


---
## 📝 curriculum
- [x] Primeiras funcoes 
- [x] Simbolos e condicionais 
- [x] Funções anonimas e lambdas
- [x] Primeiros passos com colecoes
- [x] Conhecendo mapas e threading
- [x] Map, Reduce e Filters

---

# 👩🏻‍💻 Technologies

This Project was made with:

* [Clojure](https://clojuredocs.org/)

---
## License

Copyright © 2022 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


---

<p align="center">
made with 💖 by m-carecho
</p>

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2022-05-23
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2022-05-23
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/loja/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/loja/compare/0.1.0...0.1.1


src/loja/aula3.clj
(ns loja.aula3
  (:require [loja.db :as l.db]))

(println (l.db/todos-os-pedidos))

(println (group-by :usuario (l.db/todos-os-pedidos)))

(defn minha-funcao-de-agrupamento
  [elemento]
  (println "pedido" elemento)
  (:usuario elemento))

(println (group-by minha-funcao-de-agrupamento (l.db/todos-os-pedidos)))

(println (count (vals (group-by :usuario (l.db/todos-os-pedidos)))))

(println (map count (vals (group-by :usuario (l.db/todos-os-pedidos)))))


;thread last
(->> (l.db/todos-os-pedidos)
     (group-by :usuario)
     vals
     (map count)
     println)

(defn conta-total-por-usuario
  [[usuario pedidos]]
  (count pedidos))

(->> (l.db/todos-os-pedidos)
     (group-by :usuario)
     (map conta-total-por-usuario)
     println)

(defn conta-total-por-usuario
  [[usuario pedidos]]
  [usuario (count pedidos)])

(->> (l.db/todos-os-pedidos)
     (group-by :usuario)
     (map conta-total-por-usuario)
     println)

(defn conta-total-por-usuario
  [[usuario pedidos]]
  {:usuario-id       usuario
   :total-de-pedidos (count pedidos)})


(->> (l.db/todos-os-pedidos)
     (group-by :usuario)
     (map conta-total-por-usuario)
     println)


(println "PEDIDOS")

(defn total-do-item
  [[_ detalhes]]
  (* (get detalhes :quantidade 0) (get detalhes :preco-unitario 0)))

(defn total-do-pedido
  [pedido]
  (->> pedido
       (map total-do-item)
       (reduce +)))

(defn total-dos-pedidos
       [pedidos]
       (->> pedidos
            (map :itens)
            (map total-do-pedido)
            (reduce +)))

(defn quantia-de-pedidos-e-gastos-total-por-usuario
  [[usuario pedidos]]
  {:usuario-id       usuario
   :total-de-pedidos (count pedidos)
   :preco-total    (total-dos-pedidos pedidos)})

(->> (l.db/todos-os-pedidos)
     (group-by :usuario)
     (map quantia-de-pedidos-e-gastos-total-por-usuario)
     println)












src/loja/aula5.clj
(ns loja.aula4
  (:require [loja.db :as l.db]
            [loja.logic :as l.logic]))

(println (l.db/todos-os-pedidos))

(defn resumo-por-usuario-ordenado
  [pedidos]
  (->> pedidos
       l.logic/resumo-por-usuario
       (sort-by :preco-total)
       reverse))

(let [pedidos (l.db/todos-os-pedidos)
      resumo (resumo-por-usuario-ordenado pedidos)]
  (println "Resumo" resumo)
  (println "Ordenado" (sort-by :preco-total resumo))
  (println "Ordenado ao contrario" (reverse (sort-by :preco-total resumo)))
  (println "Ordenado por id" (sort-by :usuario-id resumo))

  (println (get-in pedidos [0 :itens :mochila :quantidade])))

(let [pedidos (l.db/todos-os-pedidos)
      resumo (resumo-por-usuario-ordenado pedidos)]
  (println "Resumo" resumo)
  (println "primeiro" (first resumo))
  (println "segundo" (second resumo))
  (println "resto" (rest resumo))
  (println "total" (count resumo))
  (println "Class" (class resumo))
  (println "nth 1" (nth resumo 1))
  (println "get 1" (get resumo 1))
  (println "take" (take 2 resumo)))

(defn top-2 [resumo]
  (take 2 resumo))

(let [pedidos (l.db/todos-os-pedidos)
      resumo (resumo-por-usuario-ordenado pedidos)]
  (println "Resumo" resumo)
  (println "top-2" (top-2 resumo)))

(let [pedidos (l.db/todos-os-pedidos)
      resumo (resumo-por-usuario-ordenado pedidos)]
  (println "> 500 filter =>" (filter #(> (:preco-total %) 500) resumo))
  (println "> 500 filter empty not =>" (not (empty?  (filter #(> (:preco-total %) 500) resumo))))
  (println "> 500 some =>" (some #(> (:preco-total %) 500) resumo)))


























src/loja/aula01.clj
(ns loja.aula01)

;vetor ["daniela" "guilherme" "carlos"]
;mapa {"guilherme" 37, "paulo" 39}
;lista ligada '(1 2 3 4 5 )
;ou tem 0 ou 1 [[0 1]]
;conjunto (não repete elemento) #{}

;map - passar po rvarios elementos de um vetor fazendo algo com cada um deles e e retornando o resultado dessa execução pra cada um deles
;reduce - reduzir eles a um resultado
;filter - filtrar eles
;Utilizar a função do para rodar tudo que está dentro do if
;Utilizar a função recur para dizer que estamos fazendo uma recursão

(map println ["daniela" "guilherme" "carlos"])
;first o primeiro do vetor
;rest o resto da sequencia - rest vazio retorna uma sequencia vazia
;next - next rest vazio retorna nulo - pode ser usado pra saber quando acabou
;seq - transforma em sequencia
;reucr - a ultima coisa a se fdazer antes de retornar a funcao, tb pode ser no else
(println (first ["daniela" "guilherme" "carlos"]))
(println (rest ["daniela" "guilherme" "carlos"]))
(println (next ["daniela" "guilherme" "carlos"]))

(println "\n\n\nMeu Mapa")

;tail recursion
(defn meu-mapa
  [funcao sequencia]
  (let [primeiro (first sequencia)]
    (if (not (nil? primeiro))
      (do
        (funcao primeiro)
        (recur funcao (rest sequencia))))))

(meu-mapa println ["daniela" "guilherme" "carlos"])

src/loja/aula05.clj
(ns loja.aula05
  (:require [loja.db :as l.db]
            [loja.logic :as l.logic]))

(defn gastou-bastante? [info-do-usuairo]
  (> (:preco-total info-do-usuairo) 500))

(let [pedidos (l.db/todos-os-pedidos)
      resumo (l.logic/resumo-por-usuario pedidos)]
  (println "keep" (keep gastou-bastante? resumo)))

(println (range 10))
(println (take 2 (range 1000000000000000)))
; a sequenca nao esta sendo "gulosa" (eager)

(let [sequencia (range 1000000)]
  (println (take 2 sequencia))
  (println (take 2 sequencia)))                             ;imutabilidade, sequencia eh a mesma
; esta sendo LAZY (preguiçosa)

(defn filtro1 [x]
  (println "filtro1" x)
  x)


(defn filtro2 [x]
  (println "filtro2" x)
  x)

(->> (range 10)
     (map filtro1)
     (map filtro2)
     println)

;CHUNKS de 32
(->> (range 50)
     (map filtro1)
     (map filtro2)
     println)

(->> (range 50)
     (mapv filtro1)
     (mapv filtro2)
     println)

;Lista ligada foi 100% lazy nesse cenario
(->> '(0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0)
     (map filtro1)
     (map filtro2)
     println)


(defn str->long [valor]
  (Long/parseLong (clojure.string/replace valor #" " "")))

(println "\n -------------------------------- NOVO CLIENTE --------------------------------")

(defn novo-cliente [nome cpf email]
  {:nome  nome
   :cpf   cpf
   :email email})

(let [nome "Milena"
      cpf "123.456.789.09"
      email "milena@teste.com"]
  (println (novo-cliente nome cpf email)))


(println "\n -------------------------------- NOVO CARTAO --------------------------------")

(defn novo-cartao
  [numero cvv validade limite cliente]
  {:numero (str->long numero)
   :cvv (str->long cvv)
   :email   validade
   :limite  (bigdec limite)
   :cliente cliente})

(let [numero "1111 2222 3333 4444"
      cvv 555
      validade "2022-09"
      limite 5.000
      ;cliente "114.918.436-16"
      cliente "000.111.222-33"
      ]
  (println (novo-cartao numero cvv validade limite cliente)))



(println "\n -------------------------------- NOVA COMPRA --------------------------------")

(defn nova-compra
  [data valor estabelecimento categoria cartao]
  {:data data
   :valor (str->long valor)
   :estabelecimento estabelecimento
   :categoria categoria
   :cartao cartao})

(let [data "2022-05-19"
      valor 120
      estabelecimemto "cobasi"
      categoria "pet"
      cartao "0000 1111 2222 3333"]
  (println (nova-compra data valor estabelecimemto categoria cartao)))

src/loja/testes.clj
(ns loja.testes)

(defn str->long [valor]
  (Long/parseLong (clojure.string/replace valor #" " "")))

(println "\n -------------------------------- NOVO CLIENTE --------------------------------")

(defn novo-cliente [nome cpf email]
  {:nome  nome
   :cpf   cpf
   :email email})

(let [nome "Milena"
      cpf "123.456.789.09"
      email "milena@teste.com"]
  (println (novo-cliente nome cpf email)))

(println "\n -------------------------------- NOVO CARTAO --------------------------------")

(defn novo-cartao
  [numero cvv validade limite cliente]
  {:numero (str->long numero)
   :cvv (str->long cvv)
   :email   validade
   :limite  (bigdec limite)
   :cliente cliente})

(let [numero "0000 1111 2222 3333"
      cvv 000
      validade "2022-05"
      limite 1.000
      cliente "000.111.222-33"]
  (println (novo-cartao numero cvv validade limite cliente)))

(println "\n -------------------------------- NOVA COMPRA --------------------------------")

(defn nova-compra
  [data valor estabelecimento categoria cartao]
  {:data data
   :valor (str->long valor)
   :estabelecimento estabelecimento
   :categoria categoria
   :cartao cartao})

(let [data "2022-05-19"
      valor 120
      estabelecimemto "cobasi"
      categoria "pet"
      cartao "0000 1111 2222 3333"]
  (println (nova-compra data valor estabelecimemto categoria cartao)))

(println "\n -------------------------------- LISTAR CLIENTES, CARTÕES E COMPRAS --------------------------------")

(def clientes [{:nome "Feiticeira Escarlate" :cpf "000.111.222-33" :email "feiticeira.poderosa@vingadoras.com.br"}
               {:nome "Viúva Negra" :cpf "333.444.555-66" :email "viuva.casca.grossa@vingadoras.com.br"}
               {:nome "Hermione Granger" :cpf "666.777.888-99" :email "hermione.salvadora@hogwarts.com"}
               {:nome "Daenerys Targaryen" :cpf "999.123.456-78" :email "mae.dos.dragoes@got.com"}])

(def cartoes [{:numero "1234 1234 1234 1234" :cvv 111 :validade "2023-01" :limite 1.000 :cliente "000.111.222-33"}
              {:numero "4321 4321 4321 4321" :cvv 222 :validade "2024-02" :limite 2.000 :cliente "333.444.555-66"}
              {:numero "1598 1598 1598 1598" :cvv 333 :validade "2021-03" :limite 3.000 :cliente "666.777.888-99"}
              {:numero "3939 3939 3939 3939" :cvv 555 :validade "2026-05" :limite 5.000 :cliente "999.123.456-78"}])

(def compras [{:data "2022-01-01" :valor 129.90 :estabelecimento "Outback" :categoria "Alimentação" :cartao "1234 1234 1234 1234" }
              {:data "2022-01-02" :valor 260.00 :estabelecimento "Dentista" :categoria "Saúde" :cartao "1234 1234 1234 1234"}
              {:data "2022-02-01" :valor 20.00 :estabelecimento "Cinema" :categoria "Lazer" :cartao "1234 1234 1234 1234"}
              {:data "2022-01-10" :valor 150.00 :estabelecimento "Show" :categoria "Lazer" :cartao "4321 4321 4321 4321"}
              {:data "2022-02-10" :valor 289.99 :estabelecimento "Posto de gasolina" :categoria "Automóvel" :cartao "4321 4321 4321 4321"}
              {:data "2022-02-20" :valor 79.90 :estabelecimento "iFood" :categoria "Alimentação" :cartao "4321 4321 4321 4321"}
              {:data "2022-03-01" :valor 85.00 :estabelecimento "Alura" :categoria "Educação" :cartao "4321 4321 4321 4321"}
              {:data "2022-01-30" :valor 85.00 :estabelecimento "Alura" :categoria "Educação" :cartao "1598 1598 1598 1598"}
              {:data "2022-01-31" :valor 350.00 :estabelecimento "Tok&Stok" :categoria "Casa" :cartao "1598 1598 1598 1598"}
              {:data "2022-02-01" :valor 400.00 :estabelecimento "Leroy Merlin" :categoria "Casa" :cartao "1598 1598 1598 1598"}
              {:data "2022-03-01" :valor 50.00 :estabelecimento "Madero" :categoria "Alimentação" :cartao "6655 6655 6655 6655"}
              {:data "2022-03-01" :valor 70.00 :estabelecimento "Teatro" :categoria "Lazer" :cartao "6655 6655 6655 6655"}
              {:data "2022-03-04" :valor 250.00 :estabelecimento "Hospital" :categoria "Saúde" :cartao "6655 6655 6655 6655"}
              {:data "2022-04-10" :valor 130.00 :estabelecimento "Drogaria" :categoria "Saúde" :cartao "6655 6655 6655 6655"}
              {:data "2022-03-10" :valor 100.00 :estabelecimento "Show de pagode" :categoria "Lazer" :cartao "3939 3939 3939 3939"}
              {:data "2022-03-11" :valor 25.90 :estabelecimento "Dogão" :categoria "Alimentação" :cartao "3939 3939 3939 3939"}
              {:data "2022-03-12" :valor 215.87 :estabelecimento "Praia" :categoria "Lazer" :cartao "3939 3939 3939 3939"}
              {:data "2022-04-01" :valor 976.88 :estabelecimento "Oficina" :categoria "Automóvel" :cartao "3939 3939 3939 3939"}
              {:data "2022-04-10" :valor 85.00 :estabelecimento "Alura" :categoria "Educação" :cartao "3939 3939 3939 3939"}])


(defn todos-clientes
  []
  clientes)


(defn todos-cartoes
  []
  cartoes)


(defn todas-compras
  []
  compras)

(println "\n -------------------------------- TOTAL GASTO --------------------------------")

(defn total-gasto
  [compras]
  (->> compras (map :valor)
       (reduce +)))

(let [compras compras]
  (println (total-gasto compras)))

(println "\n -------------------------------- BUSCAR COMPRAS POR MÊS --------------------------------")

(defn mes-da-data [data]
  (second (re-matches #"\d{4}-(\d{2})-\d{2}" data)))

(defn filtra-compras
  [predicado compras]
  (vec (filter predicado compras)))

(defn busca-compras-por-mes
  [mes compras]
  (filtra-compras #(= mes (mes-da-data (:data %)))
                  compras))

(let [compras compras
      mes 3]
  (->> (busca-compras-por-mes mes compras)
       println))

(println "\n -------------------------------- BUSCAR COMPRAS POR ESTABELECIMENTO --------------------------------")

(defn busca-compras-por-estabelecimento [estabelecimento compras]
  (filtra-compras #(= estabelecimento (:estabelecimento %))
                  compras))

(let [compras compras estabelecimento "Alura"]
  (->> (busca-compras-por-estabelecimento estabelecimento compras)
       println))

(println "\n -------------------------------- CALCULAR O TOTAL DA FATURA DE UM MÊS --------------------------------")


(def total-gasto-no-mes (comp total-gasto busca-compras-por-mes))

(let [compras compras
      mes 3]
  (println (total-gasto-no-mes mes compras)))